source/common/upstream/health_discovery_service.cc

```diff
@@ -102,27 +102,49 @@ void HdsDelegate::handleFailure() {
 // TODO(lilika): Add support for the same endpoint in different clusters/ports
 envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse HdsDelegate::sendResponse() {
   envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse response;
+
   for (const auto& cluster : hds_clusters_) {
+    // Add cluster health response and set name.
+    auto* cluster_health =
+        response.mutable_endpoint_health_response()->add_cluster_endpoints_health();
+    cluster_health->set_cluster_name(cluster->info()->name());
+
+    // Iterate through all hosts in our priority set.
     for (const auto& hosts : cluster->prioritySet().hostSetsPerPriority()) {
-      for (const auto& host : hosts->hosts()) {
-        auto* endpoint = response.mutable_endpoint_health_response()->add_endpoints_health();
-        Network::Utility::addressToProtobufAddress(
-            *host->address(), *endpoint->mutable_endpoint()->mutable_address());
-        // TODO(lilika): Add support for more granular options of
-        // envoy::config::core::v3::HealthStatus
-        if (host->health() == Host::Health::Healthy) {
-          endpoint->set_health_status(envoy::config::core::v3::HEALTHY);
-        } else {
-          if (host->getActiveHealthFailureType() == Host::ActiveHealthFailureType::TIMEOUT) {
-            endpoint->set_health_status(envoy::config::core::v3::TIMEOUT);
-          } else if (host->getActiveHealthFailureType() ==
-                     Host::ActiveHealthFailureType::UNHEALTHY) {
-            endpoint->set_health_status(envoy::config::core::v3::UNHEALTHY);
-          } else if (host->getActiveHealthFailureType() == Host::ActiveHealthFailureType::UNKNOWN) {
-            endpoint->set_health_status(envoy::config::core::v3::UNHEALTHY);
+      // Get a grouping of hosts by locality.
+      for (const auto& locality_hosts : hosts->hostsPerLocality().get()) {
+        // For this locality, add the response grouping.
+        envoy::service::health::v3::LocalityEndpointsHealth* locality_health =
+            cluster_health->add_locality_endpoints_health();
+        locality_health->mutable_locality()->MergeFrom(locality_hosts[0]->locality());
+
+        // Add all hosts to this locality.
+        for (const auto& host : locality_hosts) {
+          // Add this endpoint's health status to this locality grouping.
+          auto* endpoint = locality_health->add_endpoints_health();
+          Network::Utility::addressToProtobufAddress(
+              *host->address(), *endpoint->mutable_endpoint()->mutable_address());
+          // TODO(lilika): Add support for more granular options of
+          // envoy::config::core::v3::HealthStatus
+          if (host->health() == Host::Health::Healthy) {
+            endpoint->set_health_status(envoy::config::core::v3::HEALTHY);
           } else {
-            NOT_REACHED_GCOVR_EXCL_LINE;
+            if (host->getActiveHealthFailureType() == Host::ActiveHealthFailureType::TIMEOUT) {
```

Comment:
consider a switch statement for this

Reply:
Yes, this is much cleaner than the previous implementation, since some results are shared. I went ahead and made this change while we wait for the merge. Thanks!